stix.go

```diff
@@ -968,94 +980,105 @@ func processObjects(collection *Collection, objects []json.RawMessage) error {
 
 		var obj interface{}
 
-		switch STIXType(typ.(string)) {
-		case TypeAutonomousSystem:
-			obj = &AutonomousSystem{}
-		case TypeArtifact:
-			obj = &Artifact{}
-		case TypeAttackPattern:
-			obj = &AttackPattern{}
-		case TypeCampaign:
-			obj = &Campaign{}
-		case TypeCourseOfAction:
-			obj = &CourseOfAction{}
-		case TypeDirectory:
-			obj = &Directory{}
-		case TypeDomainName:
-			obj = &DomainName{}
-		case TypeEmailAddress:
-			obj = &EmailAddress{}
-		case TypeEmailMessage:
-			obj = &EmailMessage{}
-		case TypeExtensionDefinition:
-			obj = &ExtensionDefinition{}
-		case TypeFile:
-			obj = &File{}
-		case TypeGrouping:
-			obj = &Grouping{}
-		case TypeIPv4Addr:
-			obj = &IPv4Address{}
-		case TypeIPv6Addr:
-			obj = &IPv6Address{}
-		case TypeIdentity:
-			obj = &Identity{}
-		case TypeIndicator:
-			obj = &Indicator{}
-		case TypeInfrastructure:
-			obj = &Infrastructure{}
-		case TypeIntrusionSet:
-			obj = &IntrusionSet{}
-		case TypeLanguageContent:
-			obj = &LanguageContent{}
-		case TypeLocation:
-			obj = &Location{}
-		case TypeMACAddress:
-			obj = &MACAddress{}
-		case TypeMalware:
-			obj = &Malware{}
-		case TypeMalwareAnalysis:
-			obj = &MalwareAnalysis{}
-		case TypeMarkingDefinition:
-			obj = &MarkingDefinition{}
-		case TypeMutex:
-			obj = &Mutex{}
-		case TypeNetworkTraffic:
-			obj = &NetworkTraffic{}
-		case TypeNote:
-			obj = &Note{}
-		case TypeObservedData:
-			obj = &ObservedData{}
-		case TypeOpinion:
-			obj = &Opinion{}
-		case TypeProcess:
-			obj = &Process{}
-		case TypeRegistryKey:
-			obj = &RegistryKey{}
-		case TypeRelationship:
-			obj = &Relationship{}
-		case TypeReport:
-			obj = &Report{}
-		case TypeSighting:
-			obj = &Sighting{}
-		case TypeSoftware:
-			obj = &Software{}
-		case TypeThreatActor:
-			obj = &ThreatActor{}
-		case TypeTool:
-			obj = &Tool{}
-		case TypeURL:
-			obj = &URL{}
-		case TypeUserAccount:
-			obj = &UserAccount{}
-		case TypeVulnerability:
-			obj = &Vulnerability{}
-		case TypeX509Certificate:
-			obj = &X509Certificate{}
-		default:
-			if collection.dropCustom {
-				continue
+		// Check if a custom parser should be used, if so use it.
+		if collection.customHandler != nil && collection.customHandler[STIXType(typ.(string))] != nil {
```

Comment:
Not sure if you're concerned about it but seems this would allow overriding parsing of one of the non-custom types since this happens first. When registering customHandler would it be useful to check the STIXType first?

Reply:
I added the check first to allow overriding the parsing of the standard STIX types. My reason for it is to allow building custom types using the standard STIX types. I added an [example](https://github.com/TcM1911/stix2/pull/51/files#diff-c2ade2f386c41794d5ebc57ee49b57a5fca8082e03255e5bff13977cbc061287) to the PR to show how it can be done. This has the effect that the methods such as `AttackPattern()` woun't work and the objects must be accessed via the `Get` method instead. At somepoint I will rewrite the API to use generics which should clean that up.